projects/igniteui-angular/src/lib/services/transaction/igx-hierarchical-transaction.ts

```diff
@@ -12,19 +13,52 @@ export class IgxHierarchicalTransactionService<T extends HierarchicalTransaction
         this._states.forEach((state: S, key: any) => {
             const value = mergeChanges ? this.mergeValues(state.recordRef, state.value) : state.value;
             this.clearArraysFromObject(value);
-            result.push({ id: key, parentId: state.parentId, newValue: value, type: state.type } as T);
+            result.push({ id: key, path: state.path, newValue: value, type: state.type } as T);
         });
         return result;
     }
 
     protected updateState(states: Map<any, S>, transaction: T, recordRef?: any): void {
         super.updateState(states, transaction, recordRef);
+
+        //  if transaction has no path, e.g. flat data source, get out
+        if (!transaction.path) {
+            return;
+        }
+
         const currentState = states.get(transaction.id);
-        if (currentState && transaction.type === TransactionType.ADD) {
-            currentState.parentId = transaction.parentId;
+        if (currentState) {
+            currentState.path = transaction.path;
+        }
+
+        //  if transaction has path, Hierarchical data source, and it is DELETE
+        //  type transaction for all child rows remove ADD states and update
+        //  transaction type and value of UPDATE states
+        if (transaction.type === TransactionType.DELETE) {
+            states.forEach((v: S, k: any) => {
+                if (v.path.indexOf(transaction.id) !== -1) {
+                    switch (v.type) {
+                        case TransactionType.ADD:
+                            states.delete(k);
+                            break;
+                        case TransactionType.UPDATE:
+                            states.get(k).type = TransactionType.DELETE;
+                            states.get(k).value = null;
+                    }
+                }
+            });
         }
     }
 
+    public commit(data: any[], childDataKey?: any, primaryKey?: any): void {
```

Comment:
I'd preffer if primaryKey and childDataKey parameter positions are swapped. THe primary key should generally be there and provided, child key might be optional.

Reply:
ok, done.